core/src/test/java/com/yahoo/oak/ComputeTest.java

```diff
@@ -116,14 +127,18 @@ public void testMain() throws InterruptedException {
             oak.zc().putIfAbsent(key, val);
         }
 
-        for (int i = 0; i < NUM_THREADS; i++) {
-            threads.get(i).start();
-        }
-
         latch.countDown();
 
-        for (int i = 0; i < NUM_THREADS; i++) {
-            threads.get(i).join();
+        executor.shutdown();
+        try {
+            executor.shutdown();
+            if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+                Assert.fail("should have done all the tasks in time");
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+            Assert.fail("failed to run all the tasks in the executor service");
```

Comment:
I'm not sure that `InterruptedException` means the tasked failed. You should just propagate the interrupt upwards.
According to [here](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html), the catch clause should be as follows:
```java
     } catch (InterruptedException ie) {
     // (Re-)Cancel if current thread also interrupted
     pool.shutdownNow();
     // Preserve interrupt status
     Thread.currentThread().interrupt();
   }
```

Reply:
I did this but it made me change some runnable interfaces into callables<Void> because runnable cant throw exception and callable can and it check the error by using future.Get function.